src/main/scala/com/typesafe/sbt/packager/archetypes/JavaApp.scala

```diff
@@ -62,7 +62,7 @@ object JavaAppPackaging extends AutoPlugin with JavaAppStartScript {
     scriptClasspath <<= scriptClasspathOrdering map makeRelativeClasspathNames,
     bashScriptExtraDefines := Nil,
     bashScriptConfigLocation <<= bashScriptConfigLocation ?? None,
-    bashScriptDefines <<= (Keys.mainClass in Compile, scriptClasspath, bashScriptExtraDefines, bashScriptConfigLocation) map { (mainClass, cp, extras, config) =>
+    bashScriptDefines <<= (Keys.mainClass in (Compile, bashScriptDefines), scriptClasspath in bashScriptDefines, bashScriptExtraDefines, bashScriptConfigLocation) map { (mainClass, cp, extras, config) =>
```

Comment:
By scoping these, it means the `LauncherJarPlugin` can override it just for the `bashScript`, and still use the normal configured `mainClass in Compile`.